### PR TITLE
Add empty class support for deduction

### LINKED ISSUE #3139: Deserialization of "empty" subtype with `DEDUCTION` failed
**Describe the bug**
Using `@JsonTypeInfo(use = DEDUCTION)` to deserialize a subclass with no properties/fields causing exception.

Sometimes an empty class will be used to identifying different business types:
```java
public class StartGame {
    public PaymentType paymentType;
}

@JsonTypeInfo(use = DEDUCTION)
@JsonSubTypes({@Type(Postpaid.class), @Type(Prepaid.class)})
public interface PaymentType {
}

public class Postpaid implements PaymentType {
}

public class Prepaid implements PaymentType {
    public int billId;
}
```

**Version information**
2.12.3

**To Reproduce**
```java

@JsonTypeInfo(use = DEDUCTION)
@JsonSubTypes( {@Type(Empty.class), @Type(NonEmpty.class)})
interface Base {
}

static class Empty implements Base {
}

static class NonEmpty implements Base{
  public int i;
}


public static void main(String[] args) {
  String json = "{\"i\":1}";
  Base base = mapper.readValue(json, Base.class);

  json = "{}";

  // com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve subtype of [simple type, class Base]: Cannot deduce unique subtype of `Base` (2 candidates match)
  base = mapper.readValue(json, Base.class); 
}
```

**Expected behavior**
Deserializing to `Empty` class in the example above.

### PR DESCRIPTION
**Pull request by xJoeWoo**
Fix #3139

This PR adds the ability of deserializing an empty class to `AsPropertyTypeDeserializer`.

### CODE REVIEW

**Comment by cowtowncoder** [comment]
Quick note: due to size of changes, I think this needs to go in 2.13 branch; this may be safe change but just in case.

@drekbour WDYT?

File: src/main/java/com/fasterxml/jackson/databind/jsontype/impl/AsDeductionTypeDeserializer.java [hunk 1]
```diff
@@ -35,6 +35,8 @@
     // Bitmap of available fields in each subtype (including its parents)
     private final Map<BitSet, String> subtypeFingerprints;
 
+    private static final String EMPTY_CLASS_MARKER = "";
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/jsontype/impl/AsDeductionTypeDeserializer.java:38]: Needs Javadoc to explain use.

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/jsontype/impl/AsDeductionTypeDeserializer.java:38]: One comment: empty String is actually valid property name in JSON -- is this problematic?

**Comment by cowtowncoder** [comment]
Ok so I wish I had time to dig into this deeper to make sure but... it seems to me that since "empty" class is pretty much a special case, it might make sense to not try to shoehorn it with bogus "empty property", but rather have a separate marker to indicate type with no properties, if any (and if multiple, probably fail). And then fully separate handling of "no incoming properties" case (to match to one existing no-properties pojo, if any) from handling of "one or more properties".
It seems like this might lead to even simpler handling.
I realize I may be wrong about feasibility since I haven't tried to do this.

**Comment by drekbour** [comment]
I'll take a look over the next few days

**Comment by drekbour** [comment]
Concept:

@cowtowncoder This is another example of the impossible-to-implement "deduction by absence" - one which could be handled (with compromises) by `defaultImpl`.

*However* ... the specific case of an empty subtype matching an explicit `{}` has merit and Jackson shouldn't proscribe it.

Implementation:

@xJoeWoo
There is a lot of change here to create a sentinel field representing "no fields". I think this is missing a crucial point. In the existing code an empty class is _already_ represented by an empty fingerprint (`BitSet(0)`).

With respect, I'll submit a competing PR with a simpler implementation of the same concept

**Comment by cowtowncoder** [comment]
Thank you @drekbour  -- this makes sense. I was thinking along same lines wrt this being true special case (wrt earlier discussions on why generic absence-based selection cannot be done by simple extension).

And @xJoeWoo definitely gets credit for proposing adding this useful feature (will add in release notes once we get this done), either way.

**Comment by cowtowncoder** [comment]
Merged the alternate implementation, thank you @xJoeWoo for reporting the issue & submitting the patch!

**Comment by xJoeWoo** [comment]
@drekbour The concept of your implementation is better and neater than mine. I really like to read such brilliant codes from great projects. 

@cowtowncoder It's my pleasure to make this tool better ;). And I have to admit that I didn't realize a empty name property was acceptable in json standard until you mentioned it.

**Comment by xJoeWoo** [comment]
> Concept:
> 
> @cowtowncoder This is another example of the impossible-to-implement "deduction by absence" - one which could be handled (with compromises) by `defaultImpl`.
> 
> _However_ ... the specific case of an empty subtype matching an explicit `{}` has merit and Jackson shouldn't proscribe it.
> 
> Implementation:
> 
> @xJoeWoo
> There is a lot of change here to create a sentinel field representing "no fields". I think this is missing a crucial point. In the existing code an empty class is _already_ represented by an empty fingerprint (`BitSet(0)`).
> 
> With respect, I'll submit a competing PR with a simpler implementation of the same concept

Ah.. I walked through `buildFingerprints` method again and found the empty BitSet did created if no properties were presented. Totally no need to use an empty property name to mark an empty class. I didn't catch the simpler way.

**Comment by cowtowncoder** [comment]
@xJoeWoo np at all -- it's much easier to spot various choices if you wrote the feature (in this case, @drekbour ) :)
Once again, thank you for suggesting the improvement!

**Pull request closed without merging**